### DISCUSSION TITLE
Buffer Move Semantics

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3920] [Development]
Now that `Buffer`s are starting to take off, it's time to use them more efficiently.

There are ~40 `buf_copy()`s and ~200 `buf_strcpy()`s, where an existing string is copied.
Many of these source strings are just freed afterwards.

Following navi's ideas on IRC, how about:

```c
// Free dest's contents and move src's in, leaving src empty
void buf_move(struct Buffer *dest, struct Buffer *src);

// Free dest's contents and move the string, setting the string ptr to NULL
void buf_move_str(struct Buffer *buf_dest, const char **ptr);
```

In `buf_move_str()`, the dest `Buffer` needs to know the length of the allocation.
Using `strlen()` would be good enough.

Alternatively, we could add a `int len` param.
If we know then length, pass it, if not pass `-1` and use `strlen()` in the function.

Another possibly useful function is:
```c
// Swap the contents of first and second
void buf_swap(struct Buffer *first, struct Buffer *second);
```

Care would be needed when dealing with the `BufferPool`.

**Comment by roccoblues** [comment]
> Alternatively, we could add a int len param.
> If we know then length, pass it, if not pass -1 and use strlen() in the function.

I like the explicitness of having two functions:
```c
void buf_move_str(struct Buffer *buf_dest, const char **ptr);
void buf_move_str_n(struct Buffer *buf_dest, const char **ptr, size_t len);
```
